### PR TITLE
[Truffle] is_a?: add cached fast path and TypeError

### PR DESCRIPTION
**Pull request by pitr-ch**

### CODE REVIEW

File: truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java [hunk 1]
```diff
@@ -945,14 +949,42 @@ public boolean isANil(RubyBasicObject self, Object nil) {
...
+        protected boolean IsA(RubyClass metaClass, RubyModule module) {
+            return ModuleOperations.assignableTo(metaClass, module);
+        }
+
+        protected RubyClass getMetaClass(VirtualFrame frame, Object object) {
+            return metaClassNode.executeMetaClass(frame, object);
+        }
+
+        protected int getCacheLimit() {
+            return LookupConstantNode.getCacheLimit();
```
Comment (eregon) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java]: Maybe it is better to reference the original property (= inline the code). I think these getCacheLimit() should probably be all protected.

File: truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java [hunk 2]
```diff
@@ -945,14 +949,42 @@ public boolean isANil(RubyBasicObject self, Object nil) {
...
+        @Specialization(guards = "!isRubyModule(module)")
+        public boolean isATypError(VirtualFrame frame, Object self, Object module) {
             CompilerDirectives.transferToInterpreter();
-            // TODO(CS): fast path
-            return ModuleOperations.assignableTo(getContext().getCoreLibrary().getMetaClass(self), rubyClass);
+            throw new RaiseException(getContext().getCoreLibrary().typeError("class or module required", this));
         }
 
+        @TruffleBoundary
+        protected boolean IsA(RubyClass metaClass, RubyModule module) {
```
Comment (eregon) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java]: Per naming conventions this should be `isA`.

File: truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java [hunk 3]
```diff
@@ -945,14 +949,42 @@ public boolean isANil(RubyBasicObject self, Object nil) {
...
+        }
+
         @Specialization
-        public boolean isA(Object self, RubyModule rubyClass) {
+        public boolean isAUncached(VirtualFrame frame, Object self, RubyModule module) {
+            return IsA(getMetaClass(frame, self), module);
+        }
+
+        @Specialization(guards = "!isRubyModule(module)")
+        public boolean isATypError(VirtualFrame frame, Object self, Object module) {
```
Comment (eregon) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java]: typo `isATypError`

File: truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java [hunk 4]
```diff
@@ -945,14 +949,42 @@ public boolean isANil(RubyBasicObject self, Object nil) {
...
+                                 RubyModule module,
+                                 @Cached("getMetaClass(frame, self)") RubyClass cachedMetaClass,
+                                 @Cached("module") RubyModule cachedModule,
+                                 @Cached("isA(cachedMetaClass, cachedModule)") boolean result) {
+            return result;
+        }
+
         @Specialization
-        public boolean isA(Object self, RubyModule rubyClass) {
+        public boolean isAUncached(VirtualFrame frame, Object self, RubyModule module) {
```
Comment (nirvdrum) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java]: Minor issue, but this probably should be `isAUncached`.

Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java]: Isn't it already?

Comment (nirvdrum) [inline truffle/src/main/java/org/jruby/truffle/nodes/core/KernelNodes.java]: I guess I'm losing my mind.  I swear I read `isaUncached` and I checked a few times before commenting.  Sorry for the noice.

**Pull request merged**